src/sql-server-util/src/cdc.rs

```diff
@@ -310,60 +356,114 @@ pub enum CdcError {
 /// function.
 ///
 /// [`sys.fn_cdc_increment_lsn`](https://learn.microsoft.com/en-us/sql/relational-databases/system-functions/sys-fn-cdc-increment-lsn-transact-sql?view=sql-server-ver16)
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize, Arbitrary)]
-pub struct Lsn([u8; 10]);
+///
+/// Note: The derived impl of [`PartialOrd`] and [`Ord`] relies on the field
+/// ordering so do not change it.
+#[derive(
+    Default,
+    Copy,
+    Clone,
+    Debug,
+    Eq,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Serialize,
+    Deserialize,
+    Arbitrary,
+)]
+pub struct Lsn {
+    /// Virtual Log File sequence number.
+    vlf_id: u32,
+    /// Log block number.
+    block_id: u32,
+    /// Log record number.
+    record_id: u16,
+}
 
 impl Lsn {
-    /// Interpret the provided bytes as an [`Lsn`].
-    pub fn interpret(bytes: [u8; 10]) -> Self {
-        Lsn(bytes)
-    }
+    const SIZE: usize = 10;
 
-    /// Return the underlying byte slice for this [`Lsn`].
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_slice()
-    }
+    /// Interpret the provided bytes as an [`Lsn`].
+    pub fn try_from_bytes(bytes: &[u8]) -> Result<Self, String> {
+        if bytes.len() != Self::SIZE {
+            return Err(format!("incorrect length, expected 10 got {}", bytes.len()));
+        }
 
-    /// Returns `self` as a [`StructuredLsn`].
-    pub fn as_structured(&self) -> StructuredLsn {
-        let vlf_id: [u8; 4] = self.0[0..4].try_into().expect("known good length");
-        let block_id: [u8; 4] = self.0[4..8].try_into().expect("known good length");
-        let record_id: [u8; 2] = self.0[8..].try_into().expect("known good length");
+        let vlf_id: [u8; 4] = bytes[0..4].try_into().expect("known good length");
+        let block_id: [u8; 4] = bytes[4..8].try_into().expect("known good length");
+        let record_id: [u8; 2] = bytes[8..].try_into().expect("known good length");
 
-        StructuredLsn {
+        Ok(Lsn {
             vlf_id: u32::from_be_bytes(vlf_id),
             block_id: u32::from_be_bytes(block_id),
             record_id: u16::from_be_bytes(record_id),
-        }
+        })
+    }
+
+    /// Return the underlying byte slice for this [`Lsn`].
+    pub fn as_bytes(&self) -> [u8; 10] {
+        let mut raw: [u8; Self::SIZE] = [0; 10];
+
+        raw[0..4].copy_from_slice(&self.vlf_id.to_be_bytes());
+        raw[4..8].copy_from_slice(&self.block_id.to_be_bytes());
+        raw[8..].copy_from_slice(&self.record_id.to_be_bytes());
+
+        raw
     }
-}
 
-impl Ord for Lsn {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        self.as_structured().cmp(&other.as_structured())
+    /// Increment this [`Lsn`].
+    ///
+    /// The returned [`Lsn`] may not exist upstream yet, but it's guaranteed to
+    /// sort greater than `self`.
+    pub fn increment(self) -> Lsn {
```

Comment:
If you want a more concise impl, std has a nice function that can give you back the carry to add to the rest of the fields:

```rust
let (record_id, carry) = self.record_id.overflowing_add(1);
let (block_id, carry) = self.block_id.overflowing_add(carry as u32);
let (vlf_id, overflow) = self.vlf_id.overflowing_add(carry as u32);
assert!(!overflow, "overflowed Lsn, {self:?}");
Lsn { vlf_id, block_id, record_id) }
```

Reply:
wow way better, thank you!!